contrib/slurm/slurmexecutor.py

```diff
@@ -251,57 +283,120 @@ def run_slurm(benchmark, args, log_file):
+                    first_lines,
+                )
+                jobid_match = jobid_pattern.search(str(first_lines))
+                if jobid_match:
+                    jobid = int(jobid_match.group(1))
 
         seff_command = ["seff", str(jobid)]
         logging.debug(
             "Command to run: %s", " ".join(map(util.escape_string_shell, seff_command))
         )
-        result = subprocess.run(
-            seff_command,
-            stdout=subprocess.PIPE,
-            stderr=subprocess.STDOUT,
-        )
-
-    # Runexec would populate the first 6 lines with metadata
-    with open(log_file, "r+") as file:
-        content = file.read()
-        file.seek(0, 0)
-        empty_lines = "\n" * 6
-        file.write(empty_lines + content)
 
-    exit_code, cpu_time, wall_time, memory_usage = parse_seff(str(result.stdout))
+        def get_checked_seff_result():
+            seff_result = subprocess.run(
+                seff_command,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+            )
+            if "exit code" in str(seff_result.stdout):
+                return seff_result
+            else:
+                return None
 
-    return {
-        "walltime": wall_time,
-        "cputime": cpu_time,
-        "memory": memory_usage,
-        "exitcode": ProcessExitCode.create(value=exit_code),
-    }
+        # sometimes `seff` needs a few extra seconds to realize the task has ended
+        result = wait_for(get_checked_seff_result, 30, 2)
 
+        status, exit_code, cpu_time, wall_time, memory_usage = parse_seff(
+            str(result.stdout)
+        )
 
-exit_code_pattern = re.compile(r"exit code (\d+)")
+        if os.path.exists(exitcode_file):
+            with open(exitcode_file, "r") as f:
+                returncode = int(f.read())
+                logging.debug("Exit code in file %s: %d", exitcode_file, returncode)
+        else:
+            if status == "COMPLETED":
+                raise Exception(
+                    "Should never happen: exit code not found, but task was reported COMPLETED."
+                )
+            logging.debug("Exit code not found in file: %s", exitcode_file)
+            returncode = 0
+
+        ret = {
+            "walltime": wall_time,
+            "cputime": cpu_time,
+            "memory": memory_usage,
+            "exitcode": ProcessExitCode.create(value=returncode),
+        }
+
+        if status != "COMPLETED":
+            ret["terminationreason"] = {
+                "OUT_OF_MEMORY": "memory",
+                "OUT_OF_ME+": "memory",
+                "TIMEOUT": "cputime",
+                "ERROR": "failed",
+                "FAILED": "killed",
+                "CANCELLED": "killed",
+            }.get(status, status)
+
+            if ret["terminationreason"] == "memory":
+                ret["memory"] = memory
+            if ret["terminationreason"] == "cputime":
+                ret["cputime"] = timelimit
```

Comment:
Why this? I don't think we should blindly override this. For example, if a run takes longer than the timelimit, we still want to get the real measurement.

Reply:
This is to counteract SLURM's weird way of measuring resources. SLURM will report 0 CPU time in the case of a timeout, and often 0 memory usage in the case of an OOM event. I have no idea why. To me, this seems like the next-best thing to a proper measurement: at least in the results we can see that they hit this limit. Otherwise, we'd see 0 time being spent with tasks that hit the time limit.

Example output of a job hititng its timelimit:

```
$ seff xxxxxxx
Job ID: xxxxxxx
Cluster: komondor
User/Group: p_benbl/p_benbl
State: TIMEOUT (exit code 0)
Nodes: 1
Cores per node: 3
CPU Utilized: 00:00:00
CPU Efficiency: 0% of 00:03:09 core-walltime
Job Wall-clock time: 00:01:03
Memory Utilized: 460.39 MB
Memory Efficiency: 9.21% of 4.88 GB
```

Reply:
Hm. In this case, are you really sure it used the CPU time? Does SLURM have any other time limits, like a wall-time limit?

I am really skeptical of "inventing" measurement values, because this makes it really easy to produce wrong results, if some edge case is not considered. And that would be among the worst things a benchmarking framework could do.

Reply:
I think you're right with your point about inventing measurements. I'll remove that part of the code, and just put this in the list of limitations -- afterall, the fix for this in post-processing the results is trivial.

Reply:
I removed the offending code, and the README already included this limitation. I guess I forgot to remove that line when I put this 'solution' together.